instrumentation/kafka/src/test/java/brave/kafka/ITTracingKafka.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.Time;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import zipkin.Span;
+
+public class ITTracingKafka {
+
+    private static final String LOCALHOST = "127.0.0.1";
+    private static final String KAFKA_PORT = "9092";
+
+    private static final String TEST_TOPIC = "myTopic";
+    private static final String TEST_KEY = "foo";
+    private static final String TEST_VALUE = "bar";
+
+    private KafkaServer kafkaServer;
+    private ZkUtils zkUtils;
+
+    private Tracing consumerTracing;
+    private Tracing producerTracing;
+
+    private ConcurrentLinkedDeque<Span> consumerSpans = new ConcurrentLinkedDeque<>();
+    private ConcurrentLinkedDeque<Span> producerSpans = new ConcurrentLinkedDeque<>();
+
+    @Before
+    public void setUp() throws Exception {
+        consumerTracing = Tracing.newBuilder()
+            .reporter(consumerSpans::add)
+            .sampler(Sampler.ALWAYS_SAMPLE)
+            .build();
+        producerTracing = Tracing.newBuilder()
+            .reporter(producerSpans::add)
+            .sampler(Sampler.ALWAYS_SAMPLE)
+            .build();
+
+        initKafkaServer();
+    }
+
+    @After
+    public void closeKafkaServer() {
+        kafkaServer.shutdown();
+    }
+
+    @Test
+    public void produce_and_consume_kafka_message() throws Exception {
+        TracingProducer<String, String> tracingProducer = createTracingProducer();
+        TracingConsumer<String, String> tracingConsumer = createTracingConsumer();
+
+        Future<RecordMetadata> send = tracingProducer.send(new ProducerRecord<>(TEST_TOPIC, TEST_KEY, TEST_VALUE));
+        // Block for synchronous send
+        send.get();
+
+        ConsumerRecords<String, String> records = tracingConsumer.poll(1000);
+        // Oddly we need to poll twice to retrieve content
+        if (records.isEmpty()) {
+            records = tracingConsumer.poll(1000);
+        }
+
+        assertThat(records).hasSize(1);
+        assertThat(producerSpans).hasSize(1);
+        assertThat(consumerSpans).hasSize(1);
+
+        assertThat(Long.toHexString(consumerSpans.getFirst().traceId))
+            .isEqualTo(Long.toHexString(producerSpans.getFirst().traceId));
+
+        RecordTracing recordTracing = new RecordTracing(consumerTracing);
+        for (ConsumerRecord<String, String> record : records) {
+            brave.Span span = recordTracing.nexSpanFromRecord(record);
+            assertThat(span.context().parentId()).isEqualTo(producerSpans.getLast().traceId);
+        }
+    }
+
+    private void initKafkaServer() throws IOException {
```

Comment:
I'd pull this stuff out to another file. Look at zipkin-reporter and zipkin.. I think we have a file somewhere that can be a template for approach.